### PR TITLE
Refactor internal metrics to use a Statser

### PR DESCRIPTION
**Pull request by tiedotguy**
This touches on a whole heap of stuff, but the gist is that we currently have all these backdoory hooks, and adding more internal metrics means per-metric backdoory hooks.

This removes all of them, and replaces them with a Statser.  We still use use backdoory hooks to get the Statser where it needs to be, but then it's done.  It gives a better foundation to build internal metrics on.

It also gives us the facility to send to an alternate statsd service, which means we should retain metrics under failure-to-submit conditions.  Currently that's not implemented.  Configuring Statsers is also not implemented.

Metrics now send on their own interval, rather than being slaved to the MetricFlusher, because they may not be going to the current flusher.  Intervals are also not yet configurable.

It also splits out an internal metric namespace and internal metric tags.  If we were any other service we would have this control over our internal metrics, so I want them for this service too.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -132,7 +132,7 @@ release-race: docker-race
 release: release-normal release-race
 
 run: build
-	./build/bin/$(ARCH)/$(BINARY_NAME) --backends=stdout --verbose --flush-interval=1s
+	./build/bin/$(ARCH)/$(BINARY_NAME) --backends=stdout --verbose --flush-interval=2s
```
Comment (tiedotguy) [inline Makefile]: This is here so people potentially see m vs m.count

File: pkg/statsd/dispatcher_worker.go [hunk 2]
```diff
@@ -38,3 +41,15 @@ func (w *worker) executeProcess(cmd *processCommand) {
 	defer cmd.done() // Done with the process command
 	cmd.f(w.id, w.aggr)
 }
+
+func (w *worker) runMetrics(ctx context.Context, s statser.Statser) {
+	csw := statser.NewChannelStatsWatcher(
+		s,
+		"dispatch_aggregator",
```
Comment (tiedotguy) [inline pkg/statsd/dispatcher_worker.go]: Intent of the naming is "sender_receiver".  We'll see how that pans out if I add more channels.

File: pkg/statsd/statsd_test.go [hunk 3]
```diff
@@ -82,7 +83,20 @@ func (cb *countingBackend) Name() string {
 }
 
 func (cb *countingBackend) SendMetricsAsync(ctx context.Context, m *gostatsd.MetricMap, callback gostatsd.SendCallback) {
-	atomic.AddUint64(&cb.metrics, uint64(m.NumStats))
+	count := 0
```
Comment (tiedotguy) [inline pkg/statsd/statsd_test.go]: This does not count the same value as the original.  This counts the number of tagsets, the original counted the number of datapoints.  As such, the TestStatsdThroughput summary looks worse.

File: pkg/statser/channel_stats_watcher.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+		capacity: capacity,
+		lenFunc:  lenFunc,
+		interval: interval,
+	}
+}
+
+// Run will run a ChannelStatsWatcher in the background until the supplied context is
+// closed, or Stop is called.
+func (csw *ChannelStatsWatcher) Run(ctx context.Context) {
+	go func() {
```
Comment (ash2k) [inline pkg/statser/channel_stats_watcher.go]: Let the caller decide whether they want it in a goroutine - this is much more flexible. Caller cannot undo `go` to make it run synchronously but can always add `go` to make it asynchronous. I think this is idiomatic in Go to make apis synchronous unless there are really good reasons not to.

File: pkg/statser/channel_stats_watcher.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+		ticker := time.NewTicker(csw.interval)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				ticker.Stop()
```
Comment (ash2k) [inline pkg/statser/channel_stats_watcher.go]: This is not needed because of `defer`.

File: pkg/statser/statser.go [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+	TimingMS(name string, ms float64, tags gostatsd.Tags)
+	TimingDuration(name string, d time.Duration, tags gostatsd.Tags)
+	NewTimer(name string, tags gostatsd.Tags) *Timer
+	WithTags(tags gostatsd.Tags) Statser
+}
+
+func concatTags(a, b gostatsd.Tags) gostatsd.Tags {
+	t := make(gostatsd.Tags, 0, len(a)+len(b))
+	t = append(t, a...)
+	t = append(t, b...)
```
Comment (ash2k) [inline pkg/statser/statser.go]: Hm, is it possible/better to use `copy` here since you pre-allocate the slice anyway?

File: pkg/statser/statser.go [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+	return t
+}
+
+func copyTags(tags gostatsd.Tags) gostatsd.Tags {
+	new := make(gostatsd.Tags, len(tags))
```
Comment (ash2k) [inline pkg/statser/statser.go]: `new` is a builtin function as you know :) Better not to declare vars with reserved names. Doesn't the linter complain?

File: pkg/statser/statser_internal.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+	"context"
+	"time"
+
+	"github.com/atlassian/gostatsd"
+)
+
+// InternalHandler is an interface to dispatch metrics to.  Exists
+// to break circular dependencies.
+type InternalHandler interface {
+	DispatchMetric(ctx context.Context, m *gostatsd.Metric) error
```
Comment (ash2k) [inline pkg/statser/statser_internal.go]: I usually omit formal argument names in interface declarations when typed are more than enough to understand what the arguments are.

File: pkg/statser/statser_internal.go [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+// InternalStatser is a Statser which sends metrics to a handler
+type InternalStatser struct {
+	ctx       context.Context
```
Comment (ash2k) [inline pkg/statser/statser_internal.go]: The convention is to not store context in fields but pass it as an argument i.e. it is a "per-request thing". Or when invoking a long running function. But sometimes it is inevitable and it feels like this is one of those cases.

File: pkg/statser/statser_internal.go [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+	return NewInternalStatser(is.ctx, concatTags(is.tags, tags), is.namespace, is.hostname, is.handler)
+}
+
+func (is *InternalStatser) dispatchMetric(metric *gostatsd.Metric) {
+	// the metric is owned by this file, we can change it freely because we know its origins
+	if is.namespace != "" {
+		metric.Name = is.namespace + "." + metric.Name
+	}
+	metric.Tags = append(metric.Tags, is.tags...)
+	is.handler.DispatchMetric(is.ctx, metric)
```
Comment (ash2k) [inline pkg/statser/statser_internal.go]: It returns an error but we ignore it - linter does not complain? I understand that it is the desired behaviour, but maybe explicitly ignore the error to let the reader know that this is not a mistake but on purpose?

File: pkg/statsd/dispatcher.go [hunk 11]
```diff
@@ -48,30 +52,61 @@ func NewMetricDispatcher(numWorkers int, perWorkerBufferSize int, af AggregatorF
...
 	hash := adler32.Checksum([]byte(m.Name))
 	w := d.workers[uint16(hash%uint32(d.numWorkers))]
+
+	// Protects both the read of d.running, and that the channel remains open
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+	if !d.running {
+		return nil
+	}
+
```
Comment (tiedotguy) [inline pkg/statsd/dispatcher.go]: It does, which I'm definitely not thrilled about.  RWMutex was the best I could come up with.  If there's a better way, I'm open to it.  The problem I had is the CSW tries to submit to a closing channel, which is a data race.

I originally had an explicit stop on the CSWs, but then I need to make sure I have a backdoory path to always notify CSWs to stop in future.  I'd rather they continue to be able to "submit" but have it drop.

File: pkg/statser/channel_stats_watcher.go (hunk 5 above)
Comment (tiedotguy) [inline pkg/statser/channel_stats_watcher.go]: Derp.  Good spot.

File: pkg/statser/statser.go (hunk 6 above)
Comment (tiedotguy) [inline pkg/statser/statser.go]: Possibly.  Can I do ```copy(t[len(a):], b)``` or something?  Still getting across the exact stylings of go.

File: pkg/statser/statser.go (hunk 7 above)
Comment (tiedotguy) [inline pkg/statser/statser.go]: I didn't know, shall change.  And nope, it didn't care.

File: pkg/statser/statser_internal.go (hunk 9 above)
Comment (tiedotguy) [inline pkg/statser/statser_internal.go]: I still think there's a better way of managing it all (the ctx and the handling of the channels being closed).  I couldn't unravel it all in my head though.

File: pkg/statser/statser_internal.go (hunk 8 above)
Comment (tiedotguy) [inline pkg/statser/statser_internal.go]: I like to copy and paste :)

File: pkg/statser/channel_stats_watcher.go (hunk 4 above)
Comment (tiedotguy) [inline pkg/statser/channel_stats_watcher.go]: Done

File: pkg/statsd/dispatcher.go [hunk 12]
```diff
@@ -48,30 +52,61 @@ func NewMetricDispatcher(numWorkers int, perWorkerBufferSize int, af AggregatorF
 	}
 }
 
-// Run runs the MetricDispatcher.
-func (d *MetricDispatcher) Run(ctx context.Context, done gostatsd.Done) {
```
Comment (tiedotguy) [inline pkg/statsd/dispatcher.go]: If it's launched as a go routine, then we start pumping metrics in to it before it's got running=true.  If we move the flag to NewMetricDispatcher, then we could be accepting metrics on something we haven't started, which seems like accepting metrics on something we've stopped.

**Comment by tiedotguy** [comment]
Let's have a chat about it tomorrow, hopefully you can see a better approach for addressing the close/send race.

File: pkg/statsd/dispatcher.go (hunk 11 above)
Comment (ash2k) [inline pkg/statsd/dispatcher.go]: > The problem I had is the CSW tries to submit to a closing channel, which is a data race.

Sounds like an issue that can/should be solved by controlling the shutdown order

Comment (tiedotguy) [inline pkg/statsd/dispatcher.go]: Think I got it sorted now, see how this looks.

**Comment by tiedotguy** [comment]
tl;dr: Flusher runs.  Aggregator channels fill up.  Flusher attempts to dispatch internal metrics.  Blocks because channel full.  Deadlock because the aggregators are waiting for the flush to complete.

**Comment by ash2k** [comment]
That is one reason why flusher pulls in meta metrics when it does a flush instead of metrics being pushed via standard codepath. Hm...

**Comment by tiedotguy** [comment]
Easy enough to deal with, I'll put a buffer in InternalStatser for now.  Later I'll add dropping metrics (and counting them) when the channel is full.

I would prefer to read and drop metrics, than block reading and drop packets.  The impact can be more accurately quantified that way.

**Comment by ash2k** [comment]
I think it is better to not read from the socket if we know we cannot handle it anyway. Why spend resources reading data when you can spend them processing what's been read? That was the original behaviour (it also logged them) and I changed it. We already monitor dropped packets on the interface today?

File: pkg/statser/channel_stats_watcher.go [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+	tags     gostatsd.Tags
+	capacity int
+	lenFunc  func() int
+	interval time.Duration
+}
+
+// NewChannelStatsWatcher creates a new ChannelStatsWatcher
+func NewChannelStatsWatcher(client Statser, channelName string, tags gostatsd.Tags, capacity int, lenFunc func() int, interval time.Duration) *ChannelStatsWatcher {
+	t := gostatsd.Tags{"channel:" + channelName}
+	if tags != nil {
```
Comment (ash2k) [inline pkg/statser/channel_stats_watcher.go]: This check is not necessary

File: pkg/statser/statser.go [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+func concatTags(a, b gostatsd.Tags) gostatsd.Tags {
+	t := make(gostatsd.Tags, 0, len(a)+len(b))
+	t = append(t, a...)
+	t = append(t, b...)
+	return t
+}
+
+func copyTags(tags gostatsd.Tags) gostatsd.Tags {
```
Comment (ash2k) [inline pkg/statser/statser.go]: Make it a method on `Tags`?

File: pkg/statser/statser.go [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+	Gauge(name string, value float64, tags gostatsd.Tags)
+	Count(name string, amount float64, tags gostatsd.Tags)
+	Increment(name string, tags gostatsd.Tags)
+	TimingMS(name string, ms float64, tags gostatsd.Tags)
+	TimingDuration(name string, d time.Duration, tags gostatsd.Tags)
+	NewTimer(name string, tags gostatsd.Tags) *Timer
+	WithTags(tags gostatsd.Tags) Statser
+}
+
+func concatTags(a, b gostatsd.Tags) gostatsd.Tags {
```
Comment (ash2k) [inline pkg/statser/statser.go]: Make it a method on `Tags`?

File: pkg/statser/statser.go [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+
+	"github.com/atlassian/gostatsd"
+)
+
+// Statser is the interface for sending metrics
+type Statser interface {
+	Gauge(name string, value float64, tags gostatsd.Tags)
+	Count(name string, amount float64, tags gostatsd.Tags)
+	Increment(name string, tags gostatsd.Tags)
+	TimingMS(name string, ms float64, tags gostatsd.Tags)
```
Comment (ash2k) [inline pkg/statser/statser.go]: Can we only keep `TimingDuration()`?

File: pkg/statser/statser_internal.go [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+		ctx:       ctx,
+		wg:        wg,
+		buffer:    make(chan *gostatsd.Metric, bufferSize),
+		tags:      tags,
+		namespace: namespace,
+		hostname:  hostname,
+		handler:   handler,
+	}
+	wg.Add(1)
+	go is.run()
```
Comment (ash2k) [inline pkg/statser/statser_internal.go]: Can we please make it the job of the caller to start (and then join) the goroutine? Then things get simpler - no need to pass `sync.WaitGroup`.

File: pkg/statser/channel_stats_watcher.go (hunk 13 above)
Comment (tiedotguy) [inline pkg/statser/channel_stats_watcher.go]: Wrapped up in Tags adjustments from below comments.

File: pkg/statser/statser.go (hunk 14 above)
Comment (tiedotguy) [inline pkg/statser/statser.go]: Yep.

File: pkg/statser/statser.go (hunk 15 above)
Comment (tiedotguy) [inline pkg/statser/statser.go]: Yep.

File: pkg/statser/statser_internal.go (hunk 17 above)
Comment (tiedotguy) [inline pkg/statser/statser_internal.go]: Good catch - had that for when I was creating a new InternalStatser, before the TaggingStatser.

Comment (tiedotguy) [inline pkg/statser/statser_internal.go]: Entire WG can goes away now, we don't need shutdown because the buffer can drop.

Comment (ash2k) [inline pkg/statser/statser_internal.go]: I'll take care of ordered shutdown in a separate PR later. Not just this place, but other cleanup is necessary (was necessary before your changes).
